simple-proofs/list-algorithms.md

```diff
@@ -53,19 +120,106 @@ module LIST-ALGORITHMS
     requires RHO_1 ==K RHO [ f_0 <- < lam f_lstFree LIST_FREE_BODY RHO > ]
+```
+
+and states that the execution of the body of the loop 
+(`LIST_FREE_LOOP_BODY`) in the given initial environment (discussed shortly)
+terminates, resulting in the empty list and the empty environment.
+
+This invariant reveals a pattern that can be re-used for any other recursive
+function implemented using the `Z` combinator. First, the starting term in 
+the invariant is always the loop body, and the final term is the result of
+executing *the entire loop*, not just one of its iterations. 
+
+The initial environment has a structure specific to the `Z` combinator, whereas
+the final environment is always the empty environment. In particular, the initial 
+one extends an arbitrary environment (`RHO`) with bindings for the function itself 
+(`f_lstFree`) and the passed-in parameters (in this case, `in_lst`, cf. 
+the [list-longer](#list-longer) algorithm for an example with multiple parameters). 
+While the latter are straightforward, the former is more intricate, but always 
+follows the same structure:
+
+```
+  < lam y_0 [ x_0 x_0 y_0 ] RHO_1 [ x_0 <- < lam x_0 [ f_0 ( lam y_0 [ x_0 x_0 y_0 ] ) ] RHO_1 > ] >
+```
+
+where `RHO_1` is shorthand for
+
+```
+  RHO [ f_0 <- < lam f_lstFree LIST_FREE_BODY RHO > ]
+```
+
+in which `f_0` is bound to the body of the list-free function, to be 
+evaluated in the environment `RHO`.
+
+## List-Sum
+
+The next algorithm that we consider is the list-sum algorithm, 
+`listSum(IS)`, which returns the sum of all of the values 
+provided in the list of integers `IS`. The implementation of the algorithm
+is as follows, and encode it split into three terms, as for the list-free 
+algorithm:
+
+```
+  --LIST_SUM------------- [ 
+  |                        Z (lam f_lstSum       
+  | --LIST_SUM_BODY--------- (lam in_lst 
+  |  | --...LOOP_BODY--------- (force [ (force (builtin ifThenElse))
+  |  |  |                        [ (force (builtin nullList)) in_lst ]
+  |  |  |                        ( delay (con integer 0) )
+  |  |  |                        ( delay
+  |  |  |                          [ (builtin addInteger)
+  |  |  |                            [ (force (builtin headList)) in_lst ]
+  |  |  |                            [ f_lstSum [ (force (builtin tailList)) in_lst ] ]
+  |  |  |                          ]
+  |  |  |                        )
+  |  | ----------------------- ])
+  |  ----------------------- )
+  |                        ) 
+  ----------------------- ]
+```
+
+To express correctness of list-sum, we also have to define a predicate 
+`allInts(IS)`, which checks that a given list `IS` is a list of integers:
+
+```
+  syntax Bool ::= allInts(ConstantList) [function, functional]
+
+  rule allInts(              .ConstantList) => true                              
+  rule allInts(C:Constant, IS:ConstantList) => isInt(C) andBool allInts(IS) 
+```
+
+and a predicate, `sum(XS)`, which sums a given list of integers:
```

Comment:
I think `sum(XS)` here is a function, not a predicate.

Reply:
Yes, it is.